components/zcash_note_encryption/src/lib.rs

```diff
@@ -1,7 +1,466 @@
+    type Recipient;
+    type DiversifiedTransmissionKey;
+    type IncomingViewingKey;
+    type OutgoingViewingKey;
+    type ValueCommitment;
+    type NoteCommitment;
+    type ExtractedCommitment: Eq;
+    type Memo;
+
+    fn derive_esk(note: &Self::Note) -> Option<Self::EphemeralSecretKey>;
+
+    fn get_pk_d(note: &Self::Note) -> Self::DiversifiedTransmissionKey;
+
+    fn ka_derive_public(
+        note: &Self::Note,
+        esk: &Self::EphemeralSecretKey,
+    ) -> Self::EphemeralPublicKey;
+
+    fn ka_agree_enc(
+        esk: &Self::EphemeralSecretKey,
+        pk_d: &Self::DiversifiedTransmissionKey,
+    ) -> Self::SharedSecret;
+
+    fn ka_agree_dec(
+        ivk: &Self::IncomingViewingKey,
+        epk: &Self::EphemeralPublicKey,
+    ) -> Self::SharedSecret;
+
+    fn kdf(secret: Self::SharedSecret, epk: &Self::EphemeralPublicKey) -> Self::SymmetricKey;
+
+    // for right now, we just need `recipient` to get `d`; in the future when we
+    // can get that from a Sapling note, the recipient parameter will be able
+    // to be removed.
+    fn note_plaintext_bytes(
+        note: &Self::Note,
+        recipient: &Self::Recipient,
+        memo: &Self::Memo,
+    ) -> NotePlaintextBytes;
+
+    fn get_ock(
+        ovk: &Self::OutgoingViewingKey,
+        cv: &Self::ValueCommitment,
+        cm: &Self::NoteCommitment,
+        epk: &Self::EphemeralPublicKey,
+    ) -> OutgoingCipherKey;
+
+    fn outgoing_plaintext_bytes(
+        note: &Self::Note,
+        esk: &Self::EphemeralSecretKey,
+    ) -> OutPlaintextBytes;
+
+    fn epk_bytes(epk: &Self::EphemeralPublicKey) -> EphemeralKeyBytes;
+
+    fn check_epk_bytes<F: Fn(&Self::EphemeralSecretKey) -> NoteValidity>(
+        note: &Self::Note,
+        check: F,
+    ) -> NoteValidity;
+
+    fn extract_note_commitment(note: &Self::Note) -> Self::ExtractedCommitment;
+
+    fn parse_note_plaintext_without_memo_ivk(
+        &self,
+        ivk: &Self::IncomingViewingKey,
+        plaintext: &[u8],
+    ) -> Option<(Self::Note, Self::Recipient)>;
+
+    fn parse_note_plaintext_without_memo_ovk(
+        &self,
+        pk_d: &Self::DiversifiedTransmissionKey,
+        esk: &Self::EphemeralSecretKey,
+        epk: &Self::EphemeralPublicKey,
+        plaintext: &[u8],
+    ) -> Option<(Self::Note, Self::Recipient)>;
+
+    // &self is passed here in anticipation of future changes
+    // to memo handling where the memos may no longer be
+    // part of the note plaintext.
+    fn extract_memo(&self, plaintext: &[u8]) -> Self::Memo;
+
+    fn extract_pk_d(
+        out_plaintext: &[u8; OUT_CIPHERTEXT_SIZE],
+    ) -> Option<Self::DiversifiedTransmissionKey>;
+
+    fn extract_esk(out_plaintext: &[u8; OUT_CIPHERTEXT_SIZE]) -> Option<Self::EphemeralSecretKey>;
+}
+
+pub trait ShieldedOutput<'a, D: Domain> {
+    fn ivk(&'a self) -> &'a D::IncomingViewingKey;
+    fn epk(&'a self) -> &'a D::EphemeralPublicKey;
+    fn cmstar(&'a self) -> &'a D::ExtractedCommitment;
+}
+
```

Comment:
```suggestion
```

Reply:
Oh, this is something I'd meant to come back to and had forgotten; I had intended to replace the places where these are passed as individual parameters with a single value constrained to this type.

Reply:
I have added this sketch: https://github.com/nuttycom/librustzcash/pull/8/files

This seems like a reasonably nice change to the API to me, although there is a wart with respect to CompactOutput which is that this trait cannot be implemented for it directly because the decoding of epk and cmu can fail, so I have to add a CompactDecryptable type (https://github.com/nuttycom/librustzcash/pull/8/files#diff-3ca282041189c4431fce354de87b4f2cbea97185fb8d199b86ab800c078ba899R25) so that the CompactOutput fields can first be safely deserialized before being passed anywhere this trait is used. But that's more an indictment of protobuf than anything else; the fact that one can't insert parsing to domain types into the protobuf deserialization pipeline has always been one of my main complaints about protobuf, aside from its defaulting behavior.

Reply:
I like this design (in https://github.com/nuttycom/librustzcash/pull/8).